artiq/dashboard/experiments.py

```diff
@@ -341,6 +344,15 @@ def update_repo_rev(text):
             repo_rev.textChanged.connect(update_repo_rev)
             self.repo_rev = repo_rev
 
+        recompute_sched_options = QtWidgets.QToolButton()
```

Comment:
I think separating this from the "Recompute Arguments" button is tricky. There may be hidden dependencies in an experiment that make a partial reset of only the scheduling parameters problematic (e.g. code has changed, different devices are used and pipeline name should change. arguments have been removed/moved and limits have changed). Is there a concrete use case for only resetting the scheduling parameters?

Reply:
The use case here is the same as for the individual argument recompute - you have messed up the scheduler parameters testing something (e.g. while tweaking argument values) and now want them to go back to the defaults, but want to keep the arguments.

However the main reason I was keen on a separate scheduler recompute is to keep a logical separation from the argument recompute logic. This separation allows for easy argument plugins (e.g. the [ndscan plugin](http://github.com/OxfordIonTrapGroup/ndscan.git)).

Reply:
But then you'd want individual scheduling parameter recompute buttons (like in the analogy you draw) and the overall recompute would still recompute arguments and scheduling...

The alternative is to make the scheduling defaults an overridable method (`EnvExperiment.get_scheduling_defaults()`). That would clearly separate it, address my concern and allow your use case.

Reply:
@jordens I am not sure I understand correctly:
In your suggestion, the overrideable method would return (say) a dict of default priority/pipeline/flush? Then these default values are used when the dashboard "recompute scheduler defaults" button is clicked?
This does not seem to make the "recompute" behaviour any clearer to me.

Is your objection here linked to the individual scheduler parameter setters? If so, I assume a `set_default_scheduling()` function that sets all the defaults would alleviate this? (The logic being that one setter links to one recompute button). If this is the case, I am happy implementing this.

Reply:
Sorry for the late reply. I like the general feature. I'd suggest the following:

* Phrasing: call this "reset" of scheduling settings
* Have the global reset (the recompute all arguments button) perform a reset of all arguments and all scheduling settings.
* API: `EnvExperiment.set_default_scheduling(priority=x, pipeline_name=y, flush=z, due_date=u)` to be in line with the existing "active" API around setting/creating arguments.
* If you need resets for individual items in the scheduling settings, then let's do that in a context menu to not clutter up the always-visible part of the UI.

Reply:
@jordens I agree on all but the second point.

I am hesitant to couple the scheduler options to the "recompute all arguments" button, as this belongs to the ArgumentEditor - keeping this interface clean allows easy custom argument editors (which we are already using for our scan framework [here](https://github.com/dnadlinger/artiq/tree/ndscan)).

Then the options are a "scheduler settings reset" button outside of the argument editor, as in the below, or a context menu - my preference is the former, as this seems to fit well into the current GUI. Do you have strong feelings on this?
![image](https://user-images.githubusercontent.com/13227288/53918668-80e3e300-405f-11e9-85dc-10fbce7f184b.png)

Reply:
I'm ok with keeping it out of the argument editor. But imo that button at that location makes an already cluttered and dense interface worse. It's meaning is unclear and the placement is unintuitive.

Reply:
I do see what you mean. My worry with a context menu is that many people would never notice it existed.

Another option is to shrink the submit button vertically, and add a "reset scheduler settings" button below "terminate instances", but I feel this gives more prominence than is desired to this (probably infrequently used) button. I guess context menu it is...
![image](https://user-images.githubusercontent.com/13227288/53920138-11242700-4064-11e9-9fb3-d07247a020a6.png)

Reply:
FWIW, I think context menu is the least-bad option here (there aren't any good ones, alas) -- I think this is a rarely used feature, and also that the previous suggested draft shown in https://github.com/m-labs/artiq/pull/1262#discussion_r263166497 is, as @jordens says, too cluttered and un-intuitive as to the function.